synapse/handlers/sync.py

```diff
@@ -1291,6 +1296,50 @@ async def unread_notifs_for_room_id(
                 sync_config.user.to_string(),
             )
 
+    @cached(max_entries=50000, iterable=True)
+    async def beeper_preview_for_room_id_and_user_id(
```

Comment:
Hrm, how do we invalidate this cache when new events come into the room?

Reply:
You could use an inversion of the sql for the new item as it's on it's received into the database instead of a separate query that runs for each sync as this is currently written. This could possibly reduce complexity, but given the nature of redactions and reactions we will still probably have to do a query to find the relationships and see if they are relevant to the preview calculation. i.e., a new reaction received could trigger a DB hit for every user in the room. A carefully thought out cache structure could avoid this complexity.

Pseudocode:
```
if type(new_event) is message, encryptedmessage, redaction:
    invalidate_cache(room_id, user_id)
elif type(new_event) is reaction:
    if get_event(new_event.redacts).sender == user_id:
        invalidate_cache(room_id, user_id)
```